zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -1085,15 +1108,63 @@ where
                     )?;
                     orchard_output_meta.push((
                         Recipient::InternalAccount {
-                            receiving_account: account,
+                            receiving_account: account_id,
                             external_address: None,
-                            note: PoolType::Shielded(ShieldedProtocol::Orchard),
+                            note: output_pool,
                         },
                         change_value.value(),
                         Some(memo),
                     ))
                 }
             }
+            PoolType::Transparent => {
+                #[cfg(not(feature = "transparent-inputs"))]
+                return Err(Error::UnsupportedChangeType(output_pool));
+
+                #[cfg(feature = "transparent-inputs")]
+                {
+                    num_ephemeral_addresses = num_ephemeral_addresses
+                        .checked_add(1)
+                        .ok_or(Error::ProposalNotSupported)?;
+                }
+            }
+        }
+    }
+
+    // This reserves the ephemeral addresses even if transaction construction fails.
+    // It is not worth the complexity of being able to unreserve them, because there
+    // are few failure modes after this point that would allow us to do so.
+    #[cfg(feature = "transparent-inputs")]
+    let ephemeral_addresses = wallet_db
+        .reserve_next_n_ephemeral_addresses(account_id, num_ephemeral_addresses)
+        .map_err(Error::DataSource)?;
+
+    #[cfg(feature = "transparent-inputs")]
+    {
+        assert!(ephemeral_addresses.len() == num_ephemeral_addresses.try_into().unwrap());
+
+        let mut addrs = ephemeral_addresses.iter();
+        for change_value in proposal_step.balance().proposed_change() {
+            let output_pool = change_value.output_pool();
+            if matches!(output_pool, PoolType::Transparent) {
+                // This is intended for an ephemeral transparent output, rather than a
+                // non-ephemeral transparent change output. That is, there should be another
+                // request, although we cannot check that here.
```

Comment:
This needs to be checked, so that we don't unintentionally change the semantics of "long-term" change, i.e. change that is not spent by any subsequent step.